content/blog/2022/2022-08-03-feature-flags-in-nodejs-backend-server-with-launchdarkly.md

```diff
@@ -0,0 +1,720 @@
+		console.log( chalk.dim.italic( `Loop ${ ++loop }` ) ); 
+		logger.debug( 'Executing loop.' ); 
+		try { 
+			logger.debug( 'Checking free memory.' );
+			const memoryUsed = readMemory();
+			logger.info( `Memory used: ${ memoryUsed }%` ); 
+			if ( memoryUsed >= 50 ) { 
+				logger.warn( 'More than half of free memory has been allocated.' ); 
+			} 
+		} catch ( error ) { 
+			logger.error( `Memory could not be read: ${ error.message }` );
+		} 
+		setTimeout( executeLoop, 1000 ); 
+}
+```
+
+Next we can define the script as part of `package.json`:
+
+```json
+{
+	"scripts": {
+    	"dynamic": "node dynamic_logging.js"
+  	}
+}
+```
+
+Then we can execute the following command to run our app:
+
+```bash
+npm run dynamic
+```
+
+Finally when we run the above command it will print something like below:
+
+```bash
+info: [LaunchDarkly] Initializing stream processor to receive feature flag updates
+info: [LaunchDarkly] Opened LaunchDarkly stream connection
+Loop 1
+Switching to log-level: debug
+DEBUG: Executing loop.
+DEBUG: Checking free memory.
+INFO: Memory used: 68.2%
+WARN: More than half of free memory has been allocated.
+Loop 2
+DEBUG: Executing loop.
+DEBUG: Checking free memory.
+ERROR: Memory could not be read: IOError
+Loop 3
+DEBUG: Executing loop.
+DEBUG: Checking free memory.
+INFO: Memory used: 67.4%
+WARN: More than half of free memory has been allocated.
+```
+
+### Manage blacklisted Users or IPs on the Fly
+
+Now we can take a brief look at the JSON type flags as these feature flag values are pretty open-ended. LaunchDarkly supports a JSON type right out of the box. This allows us to pass Object and Array data structures to our application which can then be used to implement lightweight administrative and operational functionality in our web application.
+
+Let’s consider that we have a server that watches the network IPs that are trying to access the server and somehow it would like to check for blacklisted IPs and then block them from our server. So if we think about it on a very bare minimum we would need the following:
+
+- a database to store the IP address.
+- an Admin UI to update those IP addresses periodically.
+- a User Access Control around those User Interfaces.
+- a data-access layer as an abstraction to update the database.
+- finally, a REST API to integrate with the UI.
+
+All of this looks tedious and need a lot of effort to maintain just a list of IP address. Well, LaunchDarkly provides us with all of this functionality with almost no effort at all. We can just define a flexible data structure and pass those values to be fetched dynamically. In this case, we can create an Array of IP addresses as part of JSON variation. This helps us to maintain a history of IP-address configuration rather than changing the same value. A type of this feature flag would look something like:
+
+{{% image alt="JSON multivariate feature flag" src="images/posts/nodejs-backend-launchdarkly/IP_Blacklist.png" %}}
```

Comment:
Again, please reduce the screenshot to the variations section to make it bigger.